proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8404,98 +8431,57 @@ lemma refill_update_is_refill_sufficient[wp]:
-  apply non_overlapping_merge_refills_simple
-  done
-
-lemma head_insufficient_loop_nonempty_refills:
-  "head_insufficient_loop sc_ptr'
-   \<lbrace>\<lambda>s. pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr\<rbrace>"
-  (is "valid ?pre _ _")
-  apply (clarsimp simp: head_insufficient_loop_def)
-  apply (wpsimp wp: valid_whileLoop[where I="\<lambda>_. ?pre"]; fastforce?)
-  apply (wpsimp wp: non_overlapping_merge_refills_nonempty_refills)
-  done
+  by (fastforce simp: refill_head_insufficient_def vs_all_heap_simps read_sched_context_def
+                      read_refill_head_def
+               split: option.splits kernel_object.splits)
 
-lemma head_insufficient_length_at_least_two:
+lemma head_insufficient_length:
   "\<lbrakk>pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg))
              (sc_refill_cfgs_of s) sc_ptr;
-    the ((head_insufficient sc_ptr) s)\<rbrakk>
-   \<Longrightarrow> pred_map (\<lambda>cfg. 1 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr"
+    refill_head_insufficient sc_ptr s = Some True\<rbrakk>
+   \<Longrightarrow> sc_refills_sc_at (\<lambda>refills. Suc 0 < length refills) sc_ptr s"
   apply (prop_tac "\<not> pred_map (\<lambda>cfg. MIN_BUDGET \<le> r_amount (scrc_refill_hd cfg))
                               (sc_refill_cfgs_of s) sc_ptr")
    apply (fastforce dest: head_insufficient_true_imp_insufficient
                     simp: vs_all_heap_simps refills_unat_sum_def)
   apply (clarsimp simp: vs_all_heap_simps refills_unat_sum_def)
   apply (rename_tac sc n)
-  apply (case_tac "sc_refills sc"; clarsimp simp: word_le_nat_alt)
+  apply (case_tac "sc_refills sc"; clarsimp simp: word_le_nat_alt sc_at_ppred_def obj_at_def)
   done
 
-lemma non_overlapping_merge_refills_refills_unat_sum:
-  "\<lbrace>\<lambda>s. pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) \<le> unat max_time)
-                      (sc_refill_cfgs_of s) sc_ptr
-        \<and> (sc_ptr' = sc_ptr \<longrightarrow> pred_map (\<lambda>cfg. 1 < length (scrc_refills cfg))
-                                         (sc_refill_cfgs_of s) sc_ptr)\<rbrace>
-   non_overlapping_merge_refills sc_ptr'
-   \<lbrace>\<lambda>_ s. pred_map (\<lambda>cfg. refills_unat_sum (scrc_refills cfg) \<le> unat max_time)
-                   (sc_refill_cfgs_of s) sc_ptr\<rbrace>"
-  apply non_overlapping_merge_refills_simple
-  apply (clarsimp simp: refills_unat_sum_def)
-  apply (rename_tac sc n)
-  apply (case_tac "sc_refills sc"; simp?)
-  apply (rename_tac list)
-  apply (subst unat_add_lem')
-   apply (prop_tac "unat (r_amount (hd list)) \<le> sum_list (map unat (map r_amount list))")
-    apply (fastforce intro: member_le_sum_list)
-   apply (clarsimp simp: unat_minus_one_word)
-  apply (case_tac list; simp)
-  done
+method merge_nonoverlapping_head_refill_simple
+  = clarsimp simp: merge_nonoverlapping_head_refill_def refill_pop_head_def merge_refill_def
```

Comment:
again consider `local_method`, I think this is only used once

Reply:
This is used 10 times, and I think the other `_simple` methods are used at least as much, so I think `supply_local_method` won't be suitable. I will see if I can consolidate them a bit.